src/commands/publish.rs

```diff
@@ -162,38 +235,154 @@ pub fn publish(publish_opts: PublishOpt) -> anyhow::Result<()> {
+    let client = reqwest::blocking::Client::builder()
+        .default_headers(reqwest::header::HeaderMap::default())
+        .build()
+        .unwrap();
+
+    let res = client
+        .post(url)
+        .header(reqwest::header::CONTENT_LENGTH, "0")
+        .header(reqwest::header::CONTENT_TYPE, "application/octet-stream")
+        .header("x-goog-resumable", "start");
+
+    let result = res.send().unwrap();
+
+    if result.status() != reqwest::StatusCode::from_u16(201).unwrap() {
+        return Err(anyhow!(
+            "Uploading package failed: got HTTP {:?} when uploading",
+            result.status()
+        ));
+    }
+
+    let headers = result
+        .headers()
+        .into_iter()
+        .filter_map(|(k, v)| {
+            let k = k.to_string();
+            let v = v.to_str().ok()?.to_string();
+            Some((k.to_lowercase(), v))
+        })
+        .collect::<BTreeMap<_, _>>();
+
+    let session_uri = headers.get("location").unwrap().clone();
+
+    let total = archived_data_size;
+
+    use indicatif::{ProgressBar, ProgressState, ProgressStyle};
+
+    // archive_path
+    let mut file = std::fs::OpenOptions::new()
+        .read(true)
+        .open(archive_path)
+        .map_err(|e| anyhow!("cannot open archive {}: {e}", archive_path.display()))?;
+
+    let pb = ProgressBar::new(archived_data_size);
+    pb.set_style(ProgressStyle::with_template("{spinner:.green} [{elapsed_precise}] [{wide_bar:.cyan/blue}] {bytes}/{total_bytes} ({eta})")
+    .unwrap()
+    .with_key("eta", |state: &ProgressState, w: &mut dyn Write| {
+        write!(w, "{:.1}s", state.eta().as_secs_f64()).unwrap()
+    })
+    .progress_chars("#>-"));
+
+    let chunk_size = 256 * 1024;
+    let mut file_pointer = 0;
+
+    let mut reader = std::io::BufReader::with_capacity(chunk_size, &mut file);
+
+    while let Some(chunk) = reader.fill_buf().ok().map(|s| s.to_vec()) {
+        let n = chunk.len();
+
+        if chunk.is_empty() {
+            break;
+        }
+
+        let start = file_pointer;
+        let end = file_pointer + chunk.len().saturating_sub(1);
+        let content_range = format!("bytes {start}-{end}/{total}");
+
+        let client = reqwest::blocking::Client::builder()
+            .default_headers(reqwest::header::HeaderMap::default())
+            .build()
+            .unwrap();
+
+        let res = client
+            .put(&session_uri)
+            .header(reqwest::header::CONTENT_TYPE, "application/octet-stream")
+            .header(reqwest::header::CONTENT_LENGTH, format!("{}", chunk.len()))
+            .header("Content-Range".to_string(), content_range)
+            .body(chunk.to_vec());
+
+        pb.set_position(file_pointer as u64);
+
+        let _response = res.send().map_err(|e| {
+            anyhow!(
+                "cannot send request to {session_uri} (chunk {}..{}): {e}",
+                file_pointer,
+                file_pointer + chunk_size
+            )
+        })?;
```

Comment:
This will only catch errors when sending the request, won't it (e.g. DNS issues or connection reset)? 

If we use [`Response::error_for_status()`](https://docs.rs/reqwest/latest/reqwest/blocking/struct.Response.html#method.error_for_status), it should automatically turn 404s and 500s into errors so we can bail out using `?`. 

```suggestion
        res
            .send()
            .map(|response| response.error_for_status())
            .map_err(|e| anyhow!(
               "cannot send request to {session_uri} (chunk {}..{}): {e}",
                file_pointer,
                file_pointer + chunk_size
            ))?;
```

(not sure if `rustfmt` will like that indenting, but you get the gist)